src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -0,0 +1,1036 @@
+            /// bitmap = 0, when the actual value is null).
+            /// </summary>
+            internal bool IsGuaranteedNull(int columnOrdinal)
+            {
+                if (_columnsCount == 0)
+                {
+                    // not an NBC row
+                    return false;
+                }
+
+                Debug.Assert(columnOrdinal >= 0 && columnOrdinal < _columnsCount, "Invalid column ordinal");
+
+                byte testBit = (byte)(1 << (columnOrdinal & 0x7)); // columnOrdinal & 0x7 == columnOrdinal MOD 0x7
+                byte testByte = _nullBitmap[columnOrdinal >> 3];
+                return (testBit & testByte) != 0;
+            }
+        }
+
+
+        /////////////////////
+        // General methods //
+        /////////////////////
+
+        // If this object is part of a TdsParserSessionPool, then this *must* be called inside the pool's lock
+        internal void Activate(object owner)
+        {
+            Debug.Assert(_parser.MARSOn, "Can not activate a non-MARS connection");
+            Owner = owner; // must assign an owner for reclamation to work
+            int result = Interlocked.Increment(ref _activateCount);   // must have non-zero activation count for reclamation to work too.
+            Debug.Assert(result == 1, "invalid deactivate count");
+        }
+
+        // CancelRequest - use to cancel while writing a request to the server
+        //
+        // o none of the request might have been sent to the server, simply reset the buffer,
+        //   sending attention does not hurt
+        // o the request was partially written. Send an ignore header to the server. attention is
+        //   required if the server was waiting for data (e.g. insert bulk rows)
+        // o the request was completely written out and the server started to process the request.
+        //   attention is required to have the server stop processing.
+        //
+        internal void CancelRequest()
+        {
+            ResetBuffer();    // clear out unsent buffer
+            // If the first sqlbulkcopy timeout, _outputPacketNumber may not be 1,
+            // the next sqlbulkcopy (same connection string) requires this to be 1, hence reset
+            // it here when exception happens in the first sqlbulkcopy
+            ResetPacketCounters();
+
+            // VSDD#907507, if bulkcopy write timeout happens, it already sent the attention,
+            // so no need to send it again
+            if (!_bulkCopyWriteTimeout)
+            {
+                SendAttention();
+                Parser.ProcessPendingAck(this);
+            }
+        }
+
+        public void CheckSetResetConnectionState(uint error, CallbackType callbackType)
+        {
+            // Should only be called for MARS - that is the only time we need to take
+            // the ResetConnection lock!
+
+            // It was raised in a security review by Microsoft questioning whether
```

Comment:
```suggestion
            // SQL BU DT 333026 - It was raised in a security review by Microsoft questioning whether
```
We may want to keep the code from netfx